src/main/java/uk/gov/hmcts/reform/idam/client/IdamClient.java

```diff
@@ -52,4 +55,15 @@ public String authenticateUser(String username, String password) {
         return BEARER_AUTH_TYPE + " " + tokenExchangeResponse.getAccessToken();
     }
 
+    public TokenExchangeResponse exchangeCode(ExchangeCodeRequest exchangeCodeRequest) {
+        return idamApi.exchangeCode(exchangeCodeRequest);
+    }
+
+    public Response authenticatePinUser(String pin, String clientId, String redirectUrl, String state) {
```

Comment:
why do you want a raw http response?
don't you want a string with the auth token in it? or the user details?

Reply:
The API (Swagger) doesn't define what comes back, and in our tests we only check it is a `200`. So it's not clear what to expect here...

Reply:
either check it or return void then I would say.

Reply:
But then the client can't check the status code..? This fits out existing tests and uses the actual Feign API...

Reply:
it's a separation of concerns.

what does your app need?
does it need to know what's in the response?
does it need to know it was successful?
does it need something specific from the response?

from an API standpoint I would only return what's needed, adding additional api surface as required but not ahead of time. 

in this case, the method returning successfully means it passed (you can do the check in the method)

Reply:
The response is the following:
```
< HTTP/1.1 302
< Cache-Control: no-store
< Pragma: no-cache
< Accept-Ranges: bytes
< Location: https://div-pfe-aat.service.core-compute-aat.internal/authenticated?code=6394fe23-cd0c-4634-905c-809573fbac59&scope=openid%20profile%20roles&iss=https%3A%2F%2Fforgerock-am.service.core-compute-idam-aat.internal%3A8443%2Fopenam%2Foauth2%2Fhmcts&client_id=divorce
< Vary: Accept-Charset, Accept-Encoding, Accept-Language, Accept
< Server: Restlet-Framework/2.3.4
< Request-Context: appId=cid-v1:0d196103-84f6-43c7-9124-898486b5197c
< X-Frame-Options: DENY
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Strict-Transport-Security: max-age=31536000 ; includeSubDomains
< Date: Fri, 16 Aug 2019 15:39:36 GMT
< Content-Length: 0
```
So we want the Location header plus the status code. It may be simpler to return the Response in this case rather than try to anticipate future usage.

Reply:
I think you want an AuthenticateUserResponse?

Maybe pull some of this code in?
https://github.com/hmcts/cmc-claim-store/blob/ba31bd390e4260d9918f2b46a99e3582980d335a/src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/idam/IdamTestService.java#L114-L135

Reply:
Ok will do, thanks Tim